node/test/Test/Pos/Client/Txp/Mode.hs

```diff
@@ -0,0 +1,291 @@
+import           Pos.Slotting                   (HasSlottingVar (..), MonadSlots (..),
+                                                 MonadSlotsData, SimpleSlottingVar,
+                                                 mkSimpleSlottingVar)
+import qualified Pos.Slotting                   as Slot
+import           Pos.Ssc.Class                  (SscBlock)
+import           Pos.Ssc.Class.Helpers          (SscHelpersClass)
+import           Pos.Ssc.GodTossing             (SscGodTossing)
+import           Pos.Util.Util                  (Some, postfixLFields)
+
+-- Need Emulation because it has instance Mockable CurrentTime
+import           Test.Pos.Block.Logic.Emulation (Emulation (..), runEmulation)
+import           Test.Pos.Block.Logic.Mode      (genSuitableStakeDistribution)
+import           Test.Pos.Client.Txp.Util       (generateAddressAux, seedSize)
+
+----------------------------------------------------------------------------
+-- TestParams
+----------------------------------------------------------------------------
+
+-- An object of this type has instance Arbitrary and is used as the source
+-- of randomness in the tests.
+data TxpTestParams = TxpTestParams
+    { _ttpGenesisContext :: !GenesisContext
+    , _ttpStartTime      :: !Microsecond
+    } deriving (Show)
+
+makeClassy ''TxpTestParams
+
+instance Arbitrary TxpTestParams where
+    arbitrary = do
+        secretKeysList <-
+            toList @(NonEmpty SecretKey) <$>
+             -- might have repetitions
+            (arbitrary `suchThat` (\l -> length l < 15))
+        let invSecretsMap = mkInvSecretsMap secretKeysList
+        let publicKeys = map toPublic (toList invSecretsMap)
+        let addresses = map makePubKeyAddressBoot publicKeys
+        let invAddrSpendingData =
+                mkInvAddrSpendingData $
+                addresses `zip` (map PubKeyASD publicKeys)
+        stakeDistribution <-
+            genSuitableStakeDistribution (fromIntegral $ length invSecretsMap)
+        let addrDistribution = [(addresses, stakeDistribution)]
+        let _ttpGenesisContext =
+                genesisContextImplicit invAddrSpendingData addrDistribution
+        let _ttpStartTime = fromMicroseconds 0
+        pure TxpTestParams {..}
+
+----------------------------------------------------------------------------
+-- Mock for TxCreateMode
+----------------------------------------------------------------------------
+
+type TxpTestMode = ReaderT TxpTestContext Emulation
+
+data TxpTestContext = TxpTestContext
+    { ttcGState       :: !GS.GStateContext
+    , ttcSystemStart  :: !Timestamp
+    , ttcSSlottingVar :: !SimpleSlottingVar
+    , ttcSlotId       :: !(Maybe SlotId)
+    -- ^ If this value is 'Just' we will return it as the current
+    -- slot. Otherwise simple slotting is used.
+    , ttcParams       :: !TxpTestParams
+    }
+
+makeLensesWith postfixLFields ''TxpTestContext
+
+instance HasCoreConstants => TxCreateMode TxpTestMode
```

Comment:
It seems this line doesn't do anything, because `TxCreateMode` is a just an alias for other constraints. I didn't even know it's possible to write such thing.
Anyway, I think it's ok to have it here, because it is helpful when code stops compiling due to a missing instance.

Reply:
It doesn't, but it acts somewhat like an assertion: it makes sure that `TxpTestMode` actually implements everything from `TxCreateMode`.